did_doc_builder/src/schema/service.rs

```diff
@@ -0,0 +1,180 @@
+use std::collections::{HashMap, HashSet};
+
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+use crate::error::DIDDocumentBuilderError;
+
+use super::{types::uri::Uri, utils::OneOrList};
+
+type ServiceTypeAlias = OneOrList<String>;
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct Service {
+    id: Uri,
+    r#type: ServiceTypeAlias,
+    service_endpoint: String,
+    #[serde(flatten)]
+    #[serde(skip_serializing_if = "HashMap::is_empty")]
+    #[serde(default)]
+    extra: HashMap<String, Value>,
+}
+
+impl Service {
+    pub fn builder(
+        id: Uri,
+        service_endpoint: String,
+    ) -> Result<ServiceBuilder, DIDDocumentBuilderError> {
+        ServiceBuilder::new(id, service_endpoint)
+    }
+
+    pub fn id(&self) -> &Uri {
+        &self.id
+    }
+
+    pub fn r#type(&self) -> &ServiceTypeAlias {
+        &self.r#type
+    }
+
+    pub fn service_endpoint(&self) -> &str {
+        self.service_endpoint.as_ref()
+    }
+
+    pub fn extra(&self, key: &str) -> Option<&Value> {
+        self.extra.get(key)
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct ServiceBuilder {
+    id: Uri,
+    r#type: HashSet<String>,
+    service_endpoint: String,
+    extra: HashMap<String, Value>,
+}
+
+impl ServiceBuilder {
+    pub fn new(id: Uri, service_endpoint: String) -> Result<Self, DIDDocumentBuilderError> {
+        if id.is_empty() {
+            return Err(DIDDocumentBuilderError::MissingField("id"));
+        }
+        if service_endpoint.is_empty() {
+            return Err(DIDDocumentBuilderError::MissingField("serviceEndpoint"));
+        }
```

Comment:
Alright, the error string makes sense then, but I still think that checking for the empty string is not really of any importance here.

The idea is that if you had something like `fn gimme_any_non_empty_string(my_str: String) {}` then I'd get why you would validate that. However, the validation is more complex here, and simply because a String is non-empty does not mean it's a valid endpoint. Trying to parse it to an `Url`, though, would validate all cases, and this check becomes redundant.